File: vital-items/src/main/java/me/vitalframework/items/VitalItem.java
package me.vitalframework.items;

import lombok.Getter;
import lombok.NonNull;
import me.vitalframework.RequiresAnnotation;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;
import org.springframework.stereotype.Component;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Used to create an item that can be interacted with.
 * This class provides a foundation for creating custom items with specific behaviors,
 * cooldowns, and interactions.
 */
@Getter
public abstract class VitalItem extends ItemStack implements RequiresAnnotation<VitalItem.Info> {
    @NonNull
    private final Map<Player, Integer> playerCooldownMap = new HashMap<>();

    private int initialCooldown = 0;

    public VitalItem() {
        final var info = getRequiredAnnotation();
        final var itemStack = new VitalItemBuilder()
                .type(info.type())
                .name(info.name())
                .amount(info.amount())
                .lore(List.of(info.lore()))
                .namespacedKey(VitalNamespacedKey.ITEM_UUID, PersistentDataType.STRING, UUID.randomUUID().toString())
                .itemFlags(List.of(info.itemFlags()))
                .unbreakable(info.unbreakable())
                .build();
        final var meta = itemStack.getItemMeta();

        if (info.enchanted()) {
            meta.addEnchant(Enchantment.LUCK, 1, true);
        }

        // TODO: cannot set item meta, "delegate" is null
        setItemMeta(meta);
        setType(itemStack.getType());
        setAmount(itemStack.getAmount());
        this.initialCooldown = info.cooldown();
    }

    public VitalItem(@NonNull ItemStack itemStack, boolean enchanted) {
        final var meta = itemStack.getItemMeta();

        if (enchanted) {
            meta.addEnchant(Enchantment.LUCK, 1, true);
        }

        // TODO: cannot set item meta, "delegate" is null
        setItemMeta(meta);
        setType(itemStack.getType());
        setAmount(itemStack.getAmount());
    }

    @Override
    public @NonNull Class<Info> requiredAnnotationType() {
        return Info.class;
    }

    public final void handleInteraction(@NonNull PlayerInteractEvent e) {
        if (!playerCooldownMap.containsKey(e.getPlayer())) {
            playerCooldownMap.put(e.getPlayer(), 0);
        }

        if (playerCooldownMap.get(e.getPlayer()) >= 1) {
            onCooldown(e);
            return;
        }

        final var action = e.getAction();

        switch (action) {
            case Action.LEFT_CLICK_AIR, Action.LEFT_CLICK_BLOCK -> onLeftClick(e);
            default -> onRightClick(e);
        }

        playerCooldownMap.put(e.getPlayer(), initialCooldown);
    }

    /**
     * Checks if this item is enchanted.
     */
    public final boolean isEnchanted() {
        return getItemMeta() == null || !getItemMeta().getEnchants().isEmpty();
    }

    /**
     * Gets the current cooldown of the given player.
     */
    public int getCooldown(@NonNull Player player) {
        return playerCooldownMap.getOrDefault(player, 0);
    }

    /**
     * Called when this item has been left-clicked.
     */
    public void onLeftClick(@NonNull PlayerInteractEvent e) {

    }

    /**
     * Called when this item has been right-clicked.
     */
    public void onRightClick(@NonNull PlayerInteractEvent e) {

    }

    /**
     * Called when this item has been left or right-clicked, but the cooldown has not yet expired.
     */
    public void onCooldown(@NonNull PlayerInteractEvent e) {

    }

    /**
     * Called when the cooldown of this item expires for the specified player.
     */
    public void onCooldownExpire(@NonNull Player player) {

    }

    /**
     * Called on every cooldown timer tick for the given player.
     */
    public void onCooldownTick(@NonNull Player player) {

    }

    @Override
    public final String toString() {
        return super.toString().replace("%s x %d"
                .formatted(getType(), getAmount()), getType() + " x 1");
    }

    @Override
    public final boolean equals(Object obj) {
        if (!(obj instanceof ItemStack item)) {
            return false;
        }

        if (item.getItemMeta() == null) {
            return item.equals(this);
        }

        if (!item.getItemMeta().getPersistentDataContainer().has(VitalNamespacedKey.ITEM_UUID, PersistentDataType.STRING)) {
            String toString = toString();

            return toString.equals(item.toString().replace("%s x %d"
                    .formatted(item.getType(), item.getAmount()), item.getType() + " x 1"));
        }

        final var uuid = UUID.fromString(getItemMeta().getPersistentDataContainer().get(VitalNamespacedKey.ITEM_UUID, PersistentDataType.STRING));
        final var otherUuid = UUID.fromString(item.getItemMeta().getPersistentDataContainer().get(VitalNamespacedKey.ITEM_UUID, PersistentDataType.STRING));

        return uuid.equals(otherUuid);
    }

    /**
     * Annotation to provide information about an item that can be interacted with in the game.
     */
    @Component
    @Target(ElementType.TYPE)
    @Retention(RetentionPolicy.RUNTIME)
    public @interface Info {
        /**
         * Defines the name of the custom item stack.
         */
        @NonNull
        String name();

        /**
         * Defines an array of lore lines for the custom item stack.
         * Lore lines provide additional information about the item.
         */
        @NonNull
        String[] lore() default {};

        /**
         * Defines the amount of items in the stack (stack size).
         */
        int amount() default 1;

        /**
         * Specifies the material or type of the custom item stack.
         */
        @NonNull
        Material type();

        /**
         * Defines an array of item flags that control specific display properties of the item stack.
         * Item flags can be used to hide attributes, enchantments, and more.
         */
        @NonNull
        ItemFlag[] itemFlags() default {};

        /**
         * Defines the cooldown time (in seconds) for the custom item stack's interaction.
         * The cooldown restricts how frequently the item can be used.
         */
        int cooldown() default 0;

        /**
         * Indicates whether the custom item stack should have enchantment visual effects.
         */
        boolean enchanted() default false;

        /**
         * Indicates whether the custom item stack should be unbreakable.
         */
        boolean unbreakable() default true;
    }
}

File: vital-items/src/main/kotlin/me/vitalframework/items/VitalItem.kt
package me.vitalframework.items

import me.vitalframework.SpigotPlayer
import me.vitalframework.Vital
import me.vitalframework.VitalCoreModule.Companion.getRequiredAnnotation
import me.vitalframework.VitalHasInfo
import me.vitalframework.items.VitalItemStackBuilder.Companion.itemBuilder
import me.vitalframework.localization.VitalLocalizationModule.Spigot.t
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import org.springframework.stereotype.Component
import java.util.UUID

/**
 * Defines an interactable item within the Vital-Framework.
 *
 * ```java
 * @VitalItem.Info(
 *   name = "MyItem",
 *   type = Material.STONE
 * )
 * public class MyItem extends VitalItem {
 *   @Override
 *   public void onLeftClick(PlayerInteractEvent e) {
 *     // ...
 *   }
 *
 *   @Override
 *   public void onRightClick(PlayerInteractEvent e) {
 *     // ...
 *   }
 *
 *   @Override
 *   public void onCooldown(PlayerInteractEvent e) {
 *     // ...
 *   }
 *
 *   @Override
 *   public void onCooldownExpire(Player player) {
 *     // ...
 *   }
 *
 *   @Override
 *   public void onCooldownTick(Player player) {
 *     // ...
 *   }
 * }
 * ```
 */
open class VitalItem : VitalHasInfo {
    override val info = mutableMapOf(Info::class.java to javaClass.getRequiredAnnotation<Info>())

    /**
     * This id is required to identify a single [VitalItem] during event processing and delegation.
     */
    val uniqueId: UUID = UUID.randomUUID()

    /**
     * The cooldowns for each player of this item.
     * Once the cooldown for a player expires, the player will be removed from this [Map].
     */
    val playerCooldown = mutableMapOf<UUID, Int>()

    /**
     * Gets the [ItemStack] for the given [player].
     */
    fun getItemStack(player: SpigotPlayer) =
        itemBuilder(uniqueId) {
            // first set default values, then try to localize them
            val info = getInfo(Info::class.java)
            type = info.type
            name = info.name
            amount = info.amount
            lore = info.lore
            itemFlags = info.itemFlags
            unbreakable = info.unbreakable

            if (info.enchanted) {
                enchantments[Enchantment.FORTUNE] = 1
            }

            if (Vital.isVitalModuleEnabled("vital-localization")) {
                name = player.t(info.name)
                lore = info.lore.flatMap { player.t(it).lines() }.toTypedArray()
                afterInit = {
                    it.itemMeta =
                        it.itemMeta.apply {
                            persistentDataContainer[VitalNamespacedKey.ITEM_LOCALIZED, PersistentDataType.BOOLEAN] =
                                true
                            persistentDataContainer[VitalNamespacedKey.ITEM_LOCALIZATION_KEY, PersistentDataType.STRING] =
                                info.name
                            persistentDataContainer[VitalNamespacedKey.ITEM_LORE_LOCALIZATION_KEYS, PersistentDataType.LIST.strings()] =
                                info.lore.toList()
                        }
                }
            } else {
                afterInit = {
                    it.itemMeta =
                        it.itemMeta.apply {
                            persistentDataContainer[VitalNamespacedKey.ITEM_LOCALIZED, PersistentDataType.BOOLEAN] =
                                false
                        }
                }
            }
        }

    /**
     * Internal function to handle the given [PlayerInteractEvent] for this item.
     * This function handles all lifecycle functions of this item and the cooldown logic.
     */
    fun handleInteraction(e: PlayerInteractEvent) {
        if (!playerCooldown.containsKey(e.player.uniqueId)) {
            playerCooldown[e.player.uniqueId] = 0
        }

        if (playerCooldown[e.player.uniqueId]!! >= 1) {
            return onCooldown(e)
        }

        when (e.action) {
            Action.LEFT_CLICK_AIR, Action.LEFT_CLICK_BLOCK -> onLeftClick(e)
            else -> onRightClick(e)
        }

        val info = getInfo(Info::class.java)
        playerCooldown[e.player.uniqueId] = info.cooldown
    }

    // TODO
    // this does not work with items if the server restarts
    // the uniqueId field is volatile and will be regenerated
    override fun equals(other: Any?): Boolean {
        if (other !is ItemStack && other !is VitalItem) {
            return false
        }

        if (other is ItemStack) {
            if (other.itemMeta == null) {
                return false
            }

            val itemUuid = other.itemMeta.persistentDataContainer[VitalNamespacedKey.ITEM_UUID, PersistentDataType.STRING]
            return uniqueId == itemUuid?.let { UUID.fromString(it) }
        }

        return uniqueId == (other as VitalItem).uniqueId
    }

    /**
     * Lifecycle function; called when this item is right-clicked by a player and the player's cooldown is not active.
     */
    open fun onLeftClick(e: PlayerInteractEvent) {}

    /**
     * Lifecycle function; called when this item is left-clicked by a player and the player's cooldown is not active.
     */
    open fun onRightClick(e: PlayerInteractEvent) {}

    /**
     * Lifecycle function; called when this item is left- or right-clicked by a player and the player's cooldown is currently active.
     */
    open fun onCooldown(e: PlayerInteractEvent) {}

    /**
     * Lifecycle function; called when this item's cooldown expires for a player.
     * This function is also called when a player doesn't have this item equipped.
     */
    open fun onCooldownExpire(player: SpigotPlayer) {}

    /**
     * Lifecycle function; called when the cooldown task ticks the cooldown for this item and a given player.
     * This function is also called when a player doesn't have this item equipped.
     */
    open fun onCooldownTick(player: SpigotPlayer) {}

    override fun toString(): String = "VitalItem(uniqueId=$uniqueId, info=$info, playerCooldown=$playerCooldown)"

    /**
     * Defines the info for a [VitalItem].
     */
    @Component
    @Target(AnnotationTarget.CLASS)
    @Retention(AnnotationRetention.RUNTIME)
    annotation class Info(
        val type: Material,
        val name: String,
        val lore: Array<String> = [],
        val cooldown: Int = 0,
        val itemFlags: Array<ItemFlag> = [],
        val amount: Int = 1,
        val enchanted: Boolean = false,
        val unbreakable: Boolean = true,
    )
}
